src/Symfony/Component/Validator/Constraints/HostnameValidator.php

```diff
@@ -0,0 +1,78 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Validator\Constraints;
+
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+use Symfony\Component\Validator\Exception\UnexpectedTypeException;
+use Symfony\Component\Validator\Exception\UnexpectedValueException;
+
+/**
+ * @author Dmitrii Poddubnyi <dpoddubny@gmail.com>
+ */
+class HostnameValidator extends ConstraintValidator
+{
+    /**
+     * https://tools.ietf.org/html/rfc2606.
+     */
+    private const RESERVED_TLDS = [
+        'example',
+        'invalid',
+        'localhost',
+        'test',
+    ];
+
+    public function validate($value, Constraint $constraint)
+    {
+        if (!$constraint instanceof Hostname) {
+            throw new UnexpectedTypeException($constraint, Hostname::class);
+        }
+
+        if (null === $value || '' === $value) {
+            return;
+        }
+
+        if (!is_scalar($value) && !(\is_object($value) && method_exists($value, '__toString'))) {
+            throw new UnexpectedValueException($value, 'string');
+        }
+
+        $value = (string) $value;
+        if ('' === $value) {
+            return;
+        }
+        if (!$this->isValid($value) || ($constraint->requireTld && !$this->hasValidTld($value))) {
+            $this->context->buildViolation($constraint->message)
+                ->setParameter('{{ value }}', $this->formatValue($value))
+                ->setCode(Hostname::INVALID_HOSTNAME_ERROR)
+                ->addViolation();
+        }
+    }
+
+    private function isValid(string $domain): bool
+    {
+        return false !== filter_var($domain, FILTER_VALIDATE_DOMAIN, FILTER_FLAG_HOSTNAME);
+    }
+
+    private function hasValidTld(string $domain): bool
+    {
+        return false !== mb_strpos($domain, '.') && !$this->hasReservedTld($domain);
+    }
+
+    private function hasReservedTld(string $domain): bool
+    {
+        $parts = explode('.', $domain);
+        $tld = end($parts);
+        $tld = mb_strtolower($tld);
+
+        return \in_array($tld, self::RESERVED_TLDS, true);
```

Comment:
`return \in_array(mb_substr($domain, mb_strrpos($domain, '.') + 1), self::RESERVED_TLDS);`

Reply:
Hi @xabbuh !
`$tld` can't have `.` since it's the last element of an array exploded by dot.
Also, did you intentionally remove the strict mode from `in_array`?

Reply:
Sorry, the third argument should be kept. But IMO we should abstain from splitting into an array here.

Reply:
I see. You suggest to replace all 3 lines with one. Got it. How about strict mode?

Reply:
I see. edited :)

Reply:
Comment fixed :)